pyiron_atomistics/sphinx/parser_base.py

```diff
@@ -0,0 +1,188 @@
+# coding: utf-8
+# Copyright (c) Max-Planck-Institut für Eisenforschung GmbH - Computational Materials Design (CM) Department
+# Distributed under the terms of "New BSD License", see the LICENSE file.
+
+__author__ = "Christoph Freysoldt"
+__copyright__ = (
+    "Copyright 2023, Max-Planck-Institut für Eisenforschung GmbH - "
+    "Computational Materials Design (CM) Department"
+)
+__version__ = "1.0"
+__maintainer__ = "Christoph Freysoldt"
+__email__ = "freysoldt@mpie.de"
+__status__ = "production"
+__date__ = "Dec 8, 2023"
+
+import re
+from types import GeneratorType
+import numpy
+
+class keyword_tree_parser:
+    """
+    A base class to parse files block by block via keyword-triggered
+    parsing routines organized in a tree. Parsing routines can
+    add more levels of keyword->parse function maps. The file
+    is read line by line on demand while parsing, so large files will not clobber
+    memory.
+
+    A parser routine can either return or yield (once!) to continue parsing. If it yields,
+    the rest of the routine (after yield) will be executed when the next keyword
+    of the current or a higher level is found.
+
+    Every parser routine MUST remove the keyword from the lineview.
+
+    A typical use will be
+
+    class my_parser(keyword_tree_parser):
+        def __init__(self,file)
+            super ().__init__({
+                    "key1" : self.parse_key1,
+                    "key2" : self.parse_key2  })
+            self.parse (file)
+
+    """
+    def __init__ (self,keylevels=[]):
+        if isinstance(keylevels,dict):
+            keylevels = [ keylevels ]
+        elif not isinstance(keylevels,list):
+            raise TypeError
+        self.keylevels = keylevels
+
+    def parse(self,filename):
+        """
+        Parse a file using the current keylevels
+
+        Args:
+            filename ... the filename of the file to parse
+        Returns: nothing
+        """
+        # --- initialization
+        if len(self.keylevels) == 0:
+           raise KeyError("No parsing functions available in keylevels")
+        filehandle = open(filename)
+        self.line = filehandle.__iter__ ()
+        self.lineview = ''
+        self.filename = filename
+        self.lineno = 0
+        while True:
+            for keymap in self.keylevels:
+                for key,func in keymap.items ():
+                    if key in self.lineview:
+                        self._cleanup(keymap)
+                        res = func ()
+                        if isinstance(res,GeneratorType):
+                            res.send (None)
+                            keymap['%finalize!'] = res
+                        break
```

Comment:
I like the use of generators, but I wonder if this cannot be streamlined a bit.

If I understand correctly, `_cleanup` is called here first, because we cannot know if the `func` from a previous iteration didn't install some new `keylevels`.  Can we not require that by providing a method on the base parser that allows temporary modification of `keylevels` but also does the clean up as a context manager.  The parser functions/generators `func` can then use this manager around their `yield` statement.

```python
from contextlib import contextmanager
class keyword_tree_parser:
  ...
  @contextmanager
  def install_keylevel(self, new_keylevels):
    self.keylevels.append(new_keylevels)
    yield
    self.keylevels.pop(-1)
```

and sub classes would use this

```python
class sub(keyword_tree_parser):
  ...
  def my_parse_method(self):
    ... # some setup
    with self.install_keylevels({'cell': self.parse_cell}):
      yield
    ... # some processing
```

If we also require all keylevel functions to be generators, the hot part of the loop can then just be

```
res = func()
res.send(None)
```

The back and forth between the parser generator functions and the base class is not quite clear to me, so maybe I'm missing some steps.